### ISSUE TITLE
Cannot use `.Namespace` in `needs` anymore

### ISSUE THREAD

**Issue by pathob** [roboll/helmfile#2098]
For some reason, I cannot use `.Namespace` in a release's `needs` block anymore. This was working with Helmfile 0.141.0 and the same Helm version. Current versions:

```
$ helmfile version
helmfile version v0.143.0
$ helm version
version.BuildInfo{Version:"v3.7.0", GitCommit:"eeac83883cb4014fe60267ec6373570374ce770b", GitTreeState:"clean", GoVersion:"go1.16.8"}
```

If you try to template the following Helmfile, it will result in the following error message:

```
repositories:
  - name: aservo
    url: https://aservo.github.io/charts

environments:
  default:

templates:
  defaults: &defaults
    name: "{{`{{ .Environment.Name }}`}}-{{`{{ .Release.Labels.service }}`}}"
    namespace: "{{`{{ .Environment.Name }}`}}"

releases:
  - chart: aservo/util
    version: 0.0.1
    labels:
      service: shared-resources
    <<: *defaults
  - chart: aservo/util
    version: 0.0.1
    labels:
      service: release-resources
    <<: *defaults
    needs:
      - "{{`{{ .Namespace }}`}}-shared-resources"
```

Output:

```
$ helmfile template
Adding repo aservo https://aservo.github.io/charts
"aservo" has been added to your repositories

in ./helmfile.yaml: release(s) "default/default-release-resources" depend(s) on an undefined release "default/{{ .Namespace }}-shared-resources". Perhaps you made a typo in "needs" or forgot defining a release named "{{ .Namespace }}-shared-resources" with appropriate "namespace" and "kubeContext"?
```

With version 0.141.0, the successful output was:

```
Adding repo aservo https://aservo.github.io/charts
"aservo" has been added to your repositories

Templating release=default-shared-resources, chart=aservo/util
Templating release=default-release-resources, chart=aservo/util
```

Unfortunately, I'm not experienced enough with Go to dig into the code myself.

**Comment by mumoshu** [comment]
@pathob Hey! Thanks for reporting.

I was rereading the implementation around it and realized that “release templates” weren’t implemented for needs.
Perhaps this never worked as intended?

I remember that, in a recent version of Helmfile (somewhere between 0.141.0 and 0.143.0), I’ve made it to explicitly fail when any of needs entries were invalid. Previously, it was silently ignored when invalid.

https://github.com/roboll/helmfile/issues/2098 is my attempt to add support for relelates templates in `needs`.
But your example still doesn't work as-is.

The important point is that a release template can't depend on the result of another release template.

That is, trying to render the `namespace` using release template `namespace: "{{`{{ .Environment.Name }}`}}"` and referring to that from the need entry `- "{{`{{ .Namespace }}`}}-shared-resources"` doesn't work.

Fortunately, you can render the namespace earlier, because it's only dependent on the environment name that is available before the release template rendering phase.

I guess this is what you wanted:

```
templates:
  defaults: &defaults
    name: "{{`{{ .Environment.Name }}`}}-{{`{{ .Release.Labels.service }}`}}"
    # Previously "{{`{{ .Environment.Name }}`}}"
    namespace: "{{ .Environment.Name }}"

#snip
    needs:
     # Previously "{{`{{ .Namespace }}`}}-shared-resources". Perhaps you meant the below?
      - "{{`{{ .Release.Namespace }}`}}-shared-resources"
```

**Comment by pathob** [comment]
Thanks @mumoshu for your analysis and yes, you're actually right, I added the `--debug` flag to the `template` command and... it never worked! So that's more like a feature request then.

And you are also right that `.Release.Namespace` would be the right way to use this. I don't remember exactly why I tried to use `.Namespace` only. At some point when I was trying to make this "work" I stumbled over a message referring to `EnvironmentTemplateData` and that one only contains a namespace string...
